### ISSUE TITLE
Copy-Pasting Line numbers; Propose: use CSS "content" for .CodeMirror-linenumber

### ISSUE THREAD

**Issue by ZombieHippie** [codemirror/CodeMirror#3695]
Hello,

I believe that when someone makes a user selection over the CodeMirror wrapper, they should not receive the line numbers in their copied text.

A solution can be achieved using [the css `content` property as described in Daniel O'Connors's article](https://danoc.me/blog/css-prevent-copy/ "article describing the concept"). This is also the solution which Github uses to display linenumbers in [their different code views](https://github.com/codemirror/CodeMirror/commit/846d043e9fd88afd98f8615b5df117de72242206 "Example diff which uses css content to display linenumbers").

If anyone can advise a good entry point for patching this, I would very much appreciate it, and if this would be an appropriate addition to the [Mode Runner library](https://codemirror.net/demo/runmode.html "Mode Runner Demo").

Merci

**Comment by marijnh** [comment]
Your patches are unavailable for some reason (github won't show them). I am not sure if a `content`-based solution would be able to support our `lineNumberFormatter` option.

**Comment by ZombieHippie** [comment]
@marijnh, thanks for the feedback, I'm looking into a way around this glitch with github.

I'm looking into the `lineNumberFormatter` option.

**Comment by marijnh** [comment]
It seems to be bugging out for the whole repository and forks right now. I can still pull stuff over the git protocol, I'll have a look in that way.

**Comment by marijnh** [comment]
Going through an attribute seems like it could work. But there's still the issue of this being entirely unsupported by some of the older browsers that we support.

**Comment by ZombieHippie** [comment]
> But there's still the issue of this being entirely unsupported by some of the older browsers that we support.

May we check compatibility to determine which method to use? I'm fairly certain that if the line numbers are `TextNode`, or if they are `css:before` content, they should style the same.

**Comment by marijnh** [comment]
Sure. Checking for `someElement.style.content != null` might cover the feature check. Also please do some benchmarks (for example creating / modifying a test page full of nodes filled with `content` vs plain text nodes) to see if there's a performance impact.

**Comment by ZombieHippie** [comment]
I'll update this issue after
[ ] I have done performance tests, and
[ ] I find resources attesting to a cross platform compatibility check.

**Comment by ZombieHippie** [comment]
@marijnh Hello, do we know of any browsers that do not support css content? [CSS Content - caniuse](http://caniuse.com/#feat=css-gencontent) [W3Schools CSS Content](http://www.w3schools.com/cssref/pr_gen_content.asp)

May you suggest a browser that this would break in? So that I may verify support checks for the feature?

I have made some [modifications to the source of my gh-pages branch](https://github.com/ZombieHippie/CodeMirror/commit/e5acb8be039a5497c535f5def3664812ebecc929) so we may see the discerning differences between css content and textnode on this ["side-by-side" testing page](http://blog.colelawr.com/CodeMirror/index-painttest.html)

Testing the change on both my desktop (very up to date hardware) and my netbook which does all graphics rendering on it's inexpensive dual core processor, there is very little difference in the speed at which the page is rendered.

Do we also need to update the code [underneath the patchDisplay code](https://github.com/codemirror/CodeMirror/blob/9951761dc3e690cb958b4f11f7a84092466c9c56/lib/codemirror.js#L882)?

**Comment by marijnh** [comment]
You're right, IE has been supporting this way longer than I thought. If you have an IE8 laying around, make sure to test your solution there.
